lib/ui/painting.dart

```diff
@@ -2146,21 +2150,119 @@ Future<Codec> instantiateImageCodecFromBuffer(
   int? targetWidth,
   int? targetHeight,
   bool allowUpscaling = true,
+}) {
+  return instantiateImageCodecWithSize(
+    buffer,
+    getTargetSize: (ImageDescriptor descriptor) {
+      if (!allowUpscaling) {
+        if (targetWidth != null && targetWidth! > descriptor.width) {
+          targetWidth = descriptor.width;
+        }
+        if (targetHeight != null && targetHeight! > descriptor.height) {
+          targetHeight = descriptor.height;
+        }
+      }
+      return TargetImageSize(width: targetWidth, height: targetHeight);
+    },
+  );
+}
+
+/// Instantiates an image [Codec].
+///
+/// This method is a convenience wrapper around the [ImageDescriptor] API.
+///
+/// The [buffer] parameter is the binary image data (e.g a PNG or GIF binary data).
+/// The data can be for either static or animated images. The following image
+/// formats are supported: {@macro dart.ui.imageFormats}
+///
+/// The [buffer] will be disposed by this method once the codec has been created,
+/// so the caller must relinquish ownership of the [buffer] when they call this
+/// method.
+///
+/// The [getTargetSize] parameter, when specified, will be invoked and passed
+/// the image's [ImageDescriptor] to determine the size to decode the image to.
+/// The width and the height of the size it returns must be positive values
+/// greater than or equal to 1, or null. It is valid to return a [TargetImageSize]
+/// that specifies only one of `width` and `height` with the other remaining
+/// null, in which case the omitted dimension will be scaled to maintain the
+/// aspect ratio of the original dimensions. When both are null or omitted,
+/// the image will be decoded at its native resolution (as will be the case if
+/// the [getTargetSize] parameter is omitted).
+///
+/// Scaling the image to larger than its intrinsic size should usually be
+/// avoided, since it causes the image to use more memory than necessary.
+/// Instead, prefer scaling the [Canvas] transform.
+///
+/// The returned future can complete with an error if the image decoding has
+/// failed.
+Future<Codec> instantiateImageCodecWithSize(
+  ImmutableBuffer buffer, {
+  TargetImageSizeProducer? getTargetSize,
 }) async {
+  getTargetSize ??= (ImageDescriptor descriptor) => const TargetImageSize();
```

Comment:
Instead of using `??=` here, create a static function and make it the argument default value:

```
TargetImageSize _defaultGetTargetSize(ImageDescriptor descriptor) {
  return const TargetImageSize();
}

```
TargetImageSizeProducer getTargetSize = _defaultGetTargetSize,
```

```

Reply:
Heh - I did that at first, but then the `package:flutter/painting.dart` API wrapper has to do something like this:

```dart
  Future<ui.Codec> instantiateImageCodecWithSize(
    ui.ImmutableBuffer buffer, {
    ui.TargetImageSizeProducer? getTargetSize,
  }) {
    if (getTargetSize == null) {
      return ui.instantiateImageCodecWithSize(buffer);
    } else {
      return ui.instantiateImageCodecWithSize(buffer, getTargetSize: getTargetSize);
    }
  }
```

Either that, or the static function has to be public (or worse: copied).  After discussing with Bob, he suggested that he frequently resorts to the method I've done here.  What do you think given those tradeoffs?

Reply:
ahh yeah, that makes sense. I think the approach is fine. I would still use a static function instead of creating a new closure on each run though

Reply:
Done